Using on RP2040 - armv6 with partial Thumb7

In attempting to build for RP2040 target="thumbv6m-none-eabi", I get the errors shown below.

Inspecting some build files, out/probe1|2|3, I see errors in : 

probe0 - r11_available is false? Due to error below.
probe1 - r7_available is false due to return values not being equal? "!0 = !{i32 63}"
probe2 - r9_available is false due to return values not being equal?  "!0 = !{i32 63}"

If the above were true, I'd see --cfg r9_reserved and there would be a build panic but I don't see any of that.  All I see is --cfg is_thumb

I'm a little confused by what is happening here.

stderr: 
```
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv6m-none-eabi` target may not support the standard library
  = note: `std` is required by `probe0` because it does not declare `#![no_std]`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: cannot use register `r11`: the frame pointer (r11) cannot be used as an operand for inline asm
 --> <anon>:2:56
  |
2 | pub fn probe() { let _ = unsafe { core::arch::asm!("", out("r11") _) }; }
  |                                                        ^^^^^^^^^^^^

error: aborting due to previous error

```


```
    Running `rustc --crate-name corosensei --edition=2021 /Users/kevinpeck/.cargo/git/checkouts/corosensei-fbd4e5d7357ec9c8/0985c39/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debuginfo=2 -C debug-assertions=on -C metadata=5a185a71dee15bf6 -C extra-filename=-5a185a71dee15bf6 --out-dir /Users/kevinpeck/Desktop/RepRap-RatRigVCore3/rp2040-project-template/target/thumbv6m-none-eabi/debug/deps --target thumbv6m-none-eabi -L dependency=/Users/kevinpeck/Desktop/RepRap-RatRigVCore3/rp2040-project-template/target/thumbv6m-none-eabi/debug/deps -L dependency=/Users/kevinpeck/Desktop/RepRap-RatRigVCore3/rp2040-project-template/target/debug/deps --extern cfg_if=/Users/kevinpeck/Desktop/RepRap-RatRigVCore3/rp2040-project-template/target/thumbv6m-none-eabi/debug/deps/libcfg_if-98ec3ee8113dde9b.rmeta --extern scopeguard=/Users/kevinpeck/Desktop/RepRap-RatRigVCore3/rp2040-project-template/target/thumbv6m-none-eabi/debug/deps/libscopeguard-4bc87324e88bd5c9.rmeta --cap-lints allow -C linker=flip-link -C link-arg=--nmagic -C link-arg=-Tlink.x -C link-arg=-Tdefmt.x -C inline-threshold=5 -C no-vectorize-loops --cfg is_thumb`

error: registers must be in range r0-r7 or lr
   |
note: instantiated into assembly here
  --> <inline asm>:13:6
   |
13 | push {r11, lr}
   |      ^

error: registers must be in range r0-r7 or pc
   |
note: instantiated into assembly here
  --> <inline asm>:23:5
   |
23 | pop {r11, pc}
   |     ^

error: registers must be in range r0-r7 or lr
   |
note: instantiated into assembly here
  --> <inline asm>:39:6
   |
39 | push {r11, lr}
   |      ^

error: instruction requires: thumb2
   |
note: instantiated into assembly here
  --> <inline asm>:40:1
   |
40 | str sp, [r1, #-8]!
   | ^

error: instruction requires: thumb2
   |
note: instantiated into assembly here
  --> <inline asm>:52:1
   |
52 | adr lr, stack_init_trampoline_return + 1
   | ^

error: invalid instruction, any one of the following would fix this:
   |
note: instantiated into assembly here
  --> <inline asm>:53:1
   |
53 | ldr pc, [r1, #4]
   | ^

note: instruction requires: thumb2
   |
note: instantiated into assembly here
  --> <inline asm>:53:1
   |
53 | ldr pc, [r1, #4]
   | ^

note: operand must be a register in range [r0, r7]
   |
note: instantiated into assembly here
  --> <inline asm>:53:5
   |
53 | ldr pc, [r1, #4]
   |     ^

error: could not compile `corosensei` due to 6 previous errors

Caused by:
  process didn't exit successfully: `... (exit status: 1)```

--- a/build.rs
+++ b/build.rs
@@ -3,6 +3,7 @@
 //! We need to determine some properties of the target for inline assembly that
 //! are not otherwise easily available. Specifically:
 //! - Does the target use the ARM or Thumb instruction set by default?
+//! - Does the target support Thumb2?
 //! - Does the target use R7 or R11 for its frame pointer?
 //! - Does the target treat R9 as a reserved register?
 
@@ -36,12 +37,21 @@ fn main() {
     // The most reliable way is to check for the thumb-mode feature in
     // CARGO_CFG_TARGET_FEATURE but this is only available on nightly. As a
     // fallback we just check if the target name starts with "thumb".
-    let is_thumb = if let Ok(target_features) = env::var("CARGO_CFG_TARGET_FEATURE") {
-        target_features.split(',').any(|s| s == "thumb-mode")
+    let (is_thumb, has_thumb2) = if let Ok(target_features) = env::var("CARGO_CFG_TARGET_FEATURE") {
+        (
+            target_features.split(',').any(|s| s == "thumb-mode"),
+            target_features.split(',').any(|s| s == "thumb2"),
+        )
     } else {
-        env::var("TARGET").unwrap().starts_with("thumb")
+        // thumbv6m-none-eabi is the only thumb target that doesn't support
+        // thumb2.
+        let target = env::var("TARGET").unwrap();
+        (target.starts_with("thumb"), !target.starts_with("thumbv6"))
     };
     if is_thumb {
         autocfg::emit("is_thumb");
+    }
+    if has_thumb2 {
+        autocfg::emit("has_thumb2");
     }
 }

--- a/src/arch/arm.rs
+++ b/src/arch/arm.rs
@@ -183,7 +183,7 @@ cfg_if::cfg_if! {
 // When computing the address of a Thumb label using ADR, we need to manually
 // set the low bit of the address to ensure we get the correct address.
 cfg_if::cfg_if! {
-    if #[cfg(is_thumb)] {
+    if #[cfg(all(is_thumb, has_thumb2))] {
         macro_rules! thumb_symbol_def {
             () => {
                 ".thumb\n.thumb_func"
@@ -192,6 +192,18 @@ cfg_if::cfg_if! {
         macro_rules! thumb_symbol_adr {
             ($reg:expr, $sym:expr) => {
                 concat!("adr ", $reg, ", ", $sym, " + 1")
+            };
+        }
+    } else if #[cfg(is_thumb)] {
+        macro_rules! thumb_symbol_def {
+            () => {
+                ".thumb\n.thumb_func"
+            };
+        }
+        // Thumb1 ADR can only represent addresses that are a multiple of 4.
+        macro_rules! thumb_symbol_adr {
+            ($reg:expr, $sym:expr) => {
+                concat!("adr ", $reg, ", ", $sym, "\n", "adds ", $reg, ", ", $reg, ", #1")
             };
         }
     } else {
@@ -204,6 +216,27 @@ cfg_if::cfg_if! {
             ($reg:expr, $sym:expr) => {
                 concat!("adr ", $reg, ", ", $sym)
             };
+        }
+    }
+}
+
+// Thumb1 support: in some cases we need to use longer instruction sequences
+// when Thumb2 is not available. This doesn't apply when using the ARM ISA since
+// that supports the full set of instructions.
+cfg_if::cfg_if! {
+    if #[cfg(any(not(is_thumb), has_thumb2))] {
+        macro_rules! thumb1 {
+            ($asm:expr) => { "" };
+        }
+        macro_rules! thumb2 {
+            ($asm:expr) => { $asm }
+        }
+    } else {
+        macro_rules! thumb1 {
+            ($asm:expr) => { $asm };
+        }
+        macro_rules! thumb2 {
+            ($asm:expr) => { "" };
         }
     }
 }
@@ -241,7 +274,10 @@ global_asm!(
     concat!("push {{", fp!(), ", lr}}"),
     // Write the parent stack pointer to the parent link and adjust R1 to point
     // to the parent link.
-    "str sp, [r1, #-8]!",
+    thumb2!("str sp, [r1, #-8]!"),
+    thumb1!("mov r3, sp"),
+    thumb1!("subs r1, r1, #8"),
+    thumb1!("str r3, [r1]"),
     // Switch to our stack while popping the padding and initial PC. This also
     // sets up r2 for the third argument to the initial function.
     "adds r2, r2, #12",
@@ -294,8 +330,18 @@ global_asm!(
     concat!(".movsp ", fp!()),
     // As in the original x86_64 code, hand-write the call operation so that it
     // doesn't push an entry into the CPU's return prediction stack.
-    thumb_symbol_adr!("lr", asm_mangle!("stack_init_trampoline_return")),
-    "ldr pc, [r1, #4]",
+    thumb2!(thumb_symbol_adr!(
+        "lr",
+        asm_mangle!("stack_init_trampoline_return")
+    )),
+    thumb2!("ldr pc, [r1, #4]"),
+    thumb1!(thumb_symbol_adr!(
+        "r3",
+        asm_mangle!("stack_init_trampoline_return")
+    )),
+    thumb1!("mov lr, r3"),
+    thumb1!("ldr r3, [r1, #4]"),
+    thumb1!("bx r3"),
     // The balign here seems to be necessary otherwise LLVM's assembler
     // generates invalid offsets to the label.
     ".balign 4",
@@ -432,14 +478,18 @@ pub unsafe fn switch_yield(arg: EncodedValue, parent_link: *mut StackPointer) ->
     let ret_val;
 
     asm_may_unwind!(
-        thumb_symbol_adr!("lr", "0f"),
+        thumb2!(thumb_symbol_adr!("lr", "0f")),
+        thumb1!(thumb_symbol_adr!("r3", "0f")),
+        thumb1!("mov lr, r3"),
         concat!("push {{r6, ", fp!(), ", lr}}"),
 
-        // Save our stack pointer to X1.
+        // Save our stack pointer to R1.
         "mov r1, sp",
 
         // Get the parent stack pointer from the parent link.
-        "ldr sp, [r2]",
+        thumb2!("ldr sp, [r2]"),
+        thumb1!("ldr r3, [r2]"),
+        thumb1!("mov sp, r3"),
 
         // Pop the FP and PC to return into the parent context.
         concat!("pop {{", fp!(), ", pc}}"),
@@ -461,14 +511,19 @@ pub unsafe fn switch_yield(arg: EncodedValue, parent_link: *mut StackPointer) ->
         concat!("push {{", fp!(), ", lr}}"),
 
         // Write the parent stack pointer to the parent link.
-        "str sp, [r1, #-8]",
+        thumb2!("str sp, [r1, #-8]"),
+        thumb1!("subs r1, r1, #8"),
+        thumb1!("mov r3, sp"),
+        thumb1!("str r3, [r1]"),
 
         // Switch to our stack
         "mov sp, r2",
 
         // Pop our R6 and FP values from the stack and implicitly pop the
         // saved PC as well.
-        concat!("pop {{r6, ", fp!(), ", lr}}"),
+        thumb2!(concat!("pop {{r6, ", fp!(), ", lr}}")),
+        thumb1!(concat!("pop {{r6, ", fp!(), "}}")),
+        thumb1!("add sp, sp, #4"),
 
         // Pass the argument in R0.
         inlateout("r0") arg => ret_val,
@@ -487,7 +542,9 @@ pub unsafe fn switch_and_reset(arg: EncodedValue, parent_link: *mut StackPointer
     // comments there. Only the differences are commented.
     asm!(
         // Load the parent context's stack pointer.
-        "ldr sp, [{parent_link}]",
+        thumb2!("ldr sp, [{parent_link}]"),
+        thumb1!("ldr r3, [{parent_link}]"),
+        thumb1!("mov sp, r3"),
 
         // Pop FP and PC off the parent stack and return into the parent
         // context.
@@ -522,7 +579,9 @@ pub unsafe fn switch_and_throw(
 
     asm_may_unwind!(
         // Set up a return address.
-        thumb_symbol_adr!("lr", "0f"),
+        thumb2!(thumb_symbol_adr!("lr", "0f")),
+        thumb1!(thumb_symbol_adr!("r3", "0f")),
+        thumb1!("mov lr, r3"),
 
         // Save the registers of the parent context.
         "push {{r6}}",
@@ -536,7 +595,10 @@ pub unsafe fn switch_and_throw(
         "mov sp, r2",
 
         // Pop the coroutine registers, with the saved PC into LR.
-        concat!("pop {{r6, ", fp!(), ", lr}}"),
+        thumb2!(concat!("pop {{r6, ", fp!(), ", lr}}")),
+        thumb1!(concat!("pop {{r6, ", fp!(), "}}")),
+        thumb1!("pop {{r3}}"),
+        thumb1!("mov lr, r3"),
 
         // Simulate a call with an artificial return address so that the throw
         // function will unwind straight into the switch_and_yield() call with

